### PR TITLE
Trainer cleanup

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

- Clean and remove unsused code in the Trainer
- The `IPUTrainer` can now accept optimizers in the init and those will be converted to `poptorch.DataLoader`
- "New" tests
- Multiple bugs fixed

Fixes: #67

### CODE REVIEW

**Comment by michaelbenayoun** [comment]
Yes, you're right, will do!

**Comment by callumm-graphcore** [comment]
_(Hi, James suggested I post these question as a comment on this PR)_

I wanted to flag up something that we hit while porting Wav2Vec2. From `optimum-graphcore/optimum/graphcore/training_args.py`, we have the following argument for training:
```
logging_nan_inf_filter: str = field(default=True, metadata={"help": "Filter nan and inf losses for logging."})
```
There were two problems we had with this:
1. The default is `True`, which meant that we missed that the loss of our model was NaN the first time we ran it.
2. The type of the argument is `str`. This meant that when I tried to pass `--logging_nan_inf_filter=False`, this was interpreted as the string `"False"`. As a non-empty string, this was interpreted as a boolean `True`, and so we didn’t see the NaN loss until we passed `--logging_nan_inf_filter=""`.

With this in mind, I have some questions:
1. Why is the default `True`? Is there a good reason to hide NaN/Inf from the user?
2. Could we at least change the type of the argument to `bool`?

File: optimum/graphcore/trainer.py [hunk 1]
```diff
@@ -959,35 +853,21 @@ def train(
...
 
         if DebugOption.UNDERFLOW_OVERFLOW in self.args.debug:
-            # TODO: test if this works.
             debug_overflow = DebugUnderflowOverflow(self.model)  # noqa
 
         self.create_optimizer_and_scheduler(num_training_steps=max_steps)
 
         self.state = TrainerState()
+        if trial is not None:
+            raise ValueError("Hyperparameyer tuning is not supported by the IPUTrainer.")
```
Comment (jimypbr) [inline optimum/graphcore/trainer.py:862]: spelling hyperparameter

File: optimum/graphcore/trainer.py [hunk 2]
```diff
@@ -303,22 +266,78 @@ def __init__(
...
+            The converted poptorch optimizer.
+        """
+        first_order_type = torch.float16 if self.ipu_config.enable_half_first_order_momentum else torch.float32
+        optimizer_kwargs = {
+            "loss_scaling": self.args.loss_scaling,
+            "accum_type": first_order_type,
+            "first_order_momentum_accum_type": first_order_type,
+            "second_order_momentum_accum_type": torch.float32,
+        }
+        # TODO: disabled max_grad_norm because it make things fail, fix it.
```
Comment (jimypbr) [inline optimum/graphcore/trainer.py:295]: what's the issue with max_grad_norm?

File: optimum/graphcore/trainer.py [hunk 3]
```diff
@@ -303,22 +266,78 @@ def __init__(
...
+        max_grad_norm = self.args.max_grad_norm
+        self.args.max_grad_norm = None
         pytorch_to_poptorch_mapping = {
-            optim.SGD: poptorch.optim.SGD,
-            optim.Adam: poptorch.optim.Adam,
-            optim.AdamW: poptorch.optim.AdamW,
-            optim.RMSpop: poptorch.optim.RMSprop,
+            optim.SGD: (poptorch.optim.SGD, {"loss_scaling": self.args.loss_scaling}),
```
Comment (jimypbr) [inline optimum/graphcore/trainer.py:299]: does loss scaling get passed to the other optimizers as well?

File: optimum/graphcore/trainer.py (hunk 1 above)
Comment (michaelbenayoun) [reply optimum/graphcore/trainer.py:862]: Thanks for catching this!

File: optimum/graphcore/trainer.py (hunk 2 above)
Comment (michaelbenayoun) [reply optimum/graphcore/trainer.py:295]: It raises an error, I think we need to fix that.
